### PR TITLE
CAL-87 Normalizing NITF attributes to use the new taxonomy

### PR DESCRIPTION
**Pull request by lcrosenbu**
#### What does this PR do?
Updates the NITF Input Transformer to align to the normalized taxonomy. 
NITF fields that are not included in the normalized taxonomy have been removed.

#### Who is reviewing it?
@dcruver 
@glenhein 
@jrnorth 
@brendan-hofmann 
@coyotesqrl 

#### Committers
@jlcsmith 

#### How should this be tested?
1. Build and run Alliance. 
2. Ingest both `image` and `gmti` NITF types. (Note: there are some samples under `imaging-transformer-nitf` resources)
3. Perform a search to retrieve the ingested NITFs.
4. Verify NITF mappings are present in the metacards.

#### Any background context you want to provide?
NITF mappings are documented here:
https://codice.atlassian.net/wiki/display/DDF/NITF

#### What are the relevant tickets?
Update NITF Input Transformer to use expanded taxonomy
https://codice.atlassian.net/browse/CAL-87

_Reference Tickets:_
Implement expanded Taxonomy in Alliance
https://codice.atlassian.net/browse/CAL-77
Implement new Taxonomy in the Catalog Core API
https://codice.atlassian.net/browse/DDF-2275

#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/MetacardFactory.java [hunk 1]
```diff
@@ -47,10 +49,10 @@
 
     public Metacard createMetacard(String id) {
         Metacard metacard = new MetacardImpl(metacardType);
-        metacard.setAttribute(new AttributeImpl(Metacard.CONTENT_TYPE, MIME_TYPE.toString()));
+        metacard.setAttribute(new AttributeImpl(Media.TYPE, MIME_TYPE.toString()));
```
Comment (bdeining) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/MetacardFactory.java]: Should we still set `Metacard.CONTENT_TYPE` for backwards compatibility?

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java [hunk 2]
```diff
@@ -19,123 +19,68 @@
...
-    COMPLEXITY_LEVEL("complexityLevel",
-            "CLEVEL",
-            NitfHeader::getComplexityLevel,
-            BasicTypes.INTEGER_TYPE),
-    STANDARD_TYPE("standardType", "STYPE", NitfHeader::getStandardType),
-    ORIGINATING_STATION_ID("originatingStationId", "OSTAID", NitfHeader::getOriginatingStationId),
+    ORIGINATING_STATION_ID(Core.SOURCE_ID,
+            "OSTAID",
+            NitfHeader::getOriginatingStationId),
     FILE_DATE_AND_TIME("fileDateAndTime",
```
Comment (dcruver) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java]: Is there a normalized taxonomy attribute for this?  I would expect all of these strings to be populated by constants.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/IndexedMtirpbAttribute.java [hunk 3]
```diff
@@ -24,30 +24,10 @@
...
-    INDEXED_TARGET_GROUND_SPEED("targetGroundSpeed",
-            "TGT_SPEED",
-            tre -> GmtiTreUtility.getTreValue(tre, "TGT_SPEED")),
-    INDEXED_TARGET_RADIAL_VELOCITY("targetRadialVelocity",
-            "TGT_VEL_R",
-            tre -> GmtiTreUtility.getTreValue(tre, "TGT_VEL_R")),
-    INDEXED_TARGET_LOCATION_ACCURACY("targetLocationAccuracy",
-            "TGT_LOC_ACCY",
-            tre -> GmtiTreUtility.getTreValue(tre, "TGT_LOC_ACCY")),
     INDEXED_TARGET_LOCATION("targetLocation",
```
Comment (dcruver) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/IndexedMtirpbAttribute.java]: Is there a normalized taxonomy attribute for this? I would expect all of these strings to be populated by constants.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/MtirpbAttribute.java [hunk 4]
```diff
@@ -26,46 +27,15 @@
...
-    WIDE_AREA_MTI_FRAME_NUMBER("wideAreaMtiFrameNumber",
-            "WAMTI_FRAME_NO",
-            tre -> GmtiTreUtility.getTreValue(tre, "WAMTI_FRAME_NO")),
-    WIDE_AREA_MTI_BAR_NUMBER("wideAreaMtiBarNumber",
-            "WAMTI_BAR_NO",
-            tre -> GmtiTreUtility.getTreValue(tre, "WAMTI_BAR_NO")),
-    SCAN_DATE_AND_TIME("scanDateAndTime",
-            "DATIME",
-            tre -> GmtiTreUtility.getTreValue(tre, "DATIME")),
     AIRCRAFT_LOCATION("aircraftLocation",
```
Comment (dcruver) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/MtirpbAttribute.java]: Is there a normalized taxonomy attribute for this? I would expect all of these strings to be populated by constants.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java (hunk 2 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java]: For the fields that are not mapped directly here, the enum is still used in the code to `getAccessorFunction`. The `FILE_DATE_AND_TIME` accessor function in particular uses the `convertNitfDate` helper function in this class, so I kept it in the list.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/MetacardFactory.java (hunk 1 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/MetacardFactory.java]: Resolved.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java (hunk 2 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java]: Removed.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/IndexedMtirpbAttribute.java (hunk 3 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/IndexedMtirpbAttribute.java]: Fixed.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/MtirpbAttribute.java (hunk 4 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/MtirpbAttribute.java]: Fixed.

File: catalog/imaging/imaging-transformer-nitf/pom.xml [hunk 5]
```diff
@@ -98,13 +110,16 @@
                         <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
                         <Embed-Dependency>
                             thumbnailator,
+                            catalog-core-api,
```
Comment (bdeining) [inline catalog/imaging/imaging-transformer-nitf/pom.xml]: We shouldn't have to embed the catalog-core-api bundle.  I think alternatively you can add this maven dependency into the imaging-app as a prereq to fix runtime issues.

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/gmti/TestGmtiAttributes.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+import org.codice.imaging.nitf.core.common.NitfFormatException;
+import org.junit.Test;
+
+public class TestGmtiAttributes {
+
+    @Test
+    public void testImageAttributes() throws NitfFormatException {
+        Stream.of(AcftbAttribute.values())
+                .forEach(attribute -> assertThat(attribute.getShortName(), is(notNullValue())));
```
Comment (bdeining) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/gmti/TestGmtiAttributes.java]: You can take out `is` for these assertions

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/gmti/TestNitfGmtiTransformer.java [hunk 7]
```diff
@@ -95,75 +100,16 @@ public void testTre() throws IOException, CatalogTransformerException, NitfForma
...
                 .getValue(), is("001"));
-        assertThat(metacard.getAttribute(MtirpbAttribute.AIRCRAFT_ALTITUDE_UNITS.getAttributeDescriptor()
-                .getName())
-                .getValue(), is("m"));
-        assertThat(metacard.getAttribute(MtirpbAttribute.AIRCRAFT_HEADING.getAttributeDescriptor()
-                .getName())
-                .getValue(), is("000"));
         assertThat(metacard.getAttribute(MtirpbAttribute.AIRCRAFT_LOCATION.getAttributeDescriptor()
                 .getName())
                 .getValue(), is("+52.123456-004.123456"));
```
Comment (dcruver) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/gmti/TestNitfGmtiTransformer.java]: When I run this locally, I see this:

java.lang.AssertionError: 
Expected: is "+52.123456-004.123456"
     but: was "POINT (+52.123456 -004.123456)"
Expected :+52.123456-004.123456
Actual   :POINT (+52.123456 -004.123456)

**Comment by dcruver** [comment]
Pulled latest branch, built, ran tests, installed kar file, ingested image and GMTI nitfs.  All worked fine.

File: catalog/imaging/imaging-transformer-nitf/pom.xml (hunk 5 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/pom.xml]: Fixed.

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/gmti/TestGmtiAttributes.java (hunk 6 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/gmti/TestGmtiAttributes.java]: Fixed.

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/gmti/TestNitfGmtiTransformer.java (hunk 7 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/gmti/TestNitfGmtiTransformer.java]: Fixed.

File: catalog/imaging/imaging-transformer-nitf/pom.xml [hunk 8]
```diff
@@ -104,7 +116,9 @@
                             codice-imaging-nitf-render,
                             codice-imaging-nitf-fluent-api
                         </Embed-Dependency>
-                        <Export-Package/>
+                        <Export-Package>
+                            ddf.catalog.*
```
Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/pom.xml]: why was this needed?

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java [hunk 9]
```diff
@@ -154,29 +92,14 @@ private NitfHeaderAttribute(final String lName, final String sName,
         this.shortName = sName;
         this.longName = lName;
         this.accessorFunction = function;
-        this.attributeDescriptor = new AttributeDescriptorImpl(ATTRIBUTE_NAME_PREFIX + longName,
+        this.attributeDescriptor = new AttributeDescriptorImpl(longName,
```
Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java]: aren't all the attribute descriptors already created and available via blueprint?

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderTransformer.java [hunk 10]
```diff
@@ -37,13 +41,26 @@ public NitfSegmentsFlow transform(NitfSegmentsFlow nitfSegmentsFlow, Metacard me
...
-        metacard.setAttribute(new AttributeImpl(Metacard.CREATED, date));
+        Date date = convertNitfDate(header.getFileDateTime());
+        metacard.setAttribute(new AttributeImpl(Core.MODIFIED, date));
+        metacard.setAttribute(new AttributeImpl(Core.CREATED, date));
         metacard.setAttribute(new AttributeImpl(Metacard.EFFECTIVE, date));
+
         handleSegmentHeader(metacard, header, NitfHeaderAttribute.values());
     }
+
+    private static Date convertNitfDate(DateTime nitfDateTime) {
```
Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderTransformer.java]: might be better to use Optional here

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/AcftbAttribute.java [hunk 11]
```diff
@@ -73,7 +51,7 @@
         this.longName = longName;
         this.shortName = shortName;
         this.accessorFunction = accessorFunction;
-        this.attributeDescriptor = new AttributeDescriptorImpl(ATTRIBUTE_NAME_PREFIX + longName,
+        this.attributeDescriptor = new AttributeDescriptorImpl(longName,
```
Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/AcftbAttribute.java]: again, I don't think we need to re-create the attribute descriptors

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java [hunk 12]
```diff
@@ -67,65 +70,64 @@ public Metacard transform(NitfSegmentsFlow nitfSegmentsFlow, Metacard metacard)
...
+        tres.stream().filter(tre -> MTIRPB.equals(tre.getName().trim())).forEach(tre -> {
+            handleSegmentHeader(metacard, tre, MtirpbAttribute.values());
+
+            try {
+                List<TreGroup> targets = tre.getEntry(TARGETS).getGroups();
+
+                targets.stream().forEach(group -> handleSegmentHeader(metacard, group,
+                        IndexedMtirpbAttribute.values()));
+            } catch (NitfFormatException e) {
+                LOGGER.error(e.getMessage(), e);
```
Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java]: probably can log this at WARN.  Also you can replace e.getMessage() with something more admin friendly as the exception message will get logged with the exception.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java [hunk 13]
```diff
@@ -135,6 +137,47 @@ private String setLocation(Metacard metacard) {
...
+        Attribute aircraftLocation = metacard
+                .getAttribute(MtirpbAttribute.AIRCRAFT_LOCATION.getAttributeDescriptor().getName());
+
+        if (aircraftLocation != null
+                && StringUtils.isNotEmpty(aircraftLocation.getValue().toString())) {
+
+            String unformattedAircraftLocation = aircraftLocation.getValue().toString();
+
+            StringBuilder sb = new StringBuilder("POINT (");
+            parseLocation(sb, unformattedAircraftLocation);
```
Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java]: can we depend on the coordinate order being LONG LAT in all cases?  Is this field always decimal degrees?

Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java]: disregard - I just saw parseLocation()

File: catalog/imaging/imaging-transformer-nitf/pom.xml (hunk 8 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/pom.xml]: Refactored the dependencies and this is no longer needed. Removed.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java (hunk 9 above)
Comment (dcruver) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java]: They're available, but not directly.  There are a few choices.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/ImageAttribute.java [hunk 14]
```diff
@@ -30,189 +31,45 @@
...
-            "ISCTLN",
-            segment -> segment.getSecurityMetadata()
-                    .getSecurityControlNumber()),
-    IMAGE_SOURCE("imageSource", "ISORCE", ImageSegment::getImageSource),
-    NUMBER_OF_SIGNIFICANT_ROWS_IN_IMAGE("numberOfSignificantRowsInImage",
+    TARGET_IDENTIFIER(Isr.TARGET_ID,
+            "TGTID",
+            segment -> getTargetId(segment)),
+    IMAGE_IDENTIFIER_2(Isr.IMAGE_ID,
+            "IID2",
```
Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/ImageAttribute.java]: Not for anyone wondering, it appears by convention that IID2 is the field where the "normal" 40 character image ID is stored.  IID2 appears to be used for other identifying information which has less value to end users.

**Comment by jlcsmith** [comment]
Built and tested.  Mappings look correct to me. Other than my few comments 👍

**Comment by kcwire** [comment]
👍
